### ISSUE TITLE
add ability to get videos from test runs

### ISSUE THREAD

**Issue by tmc** [ysmood/rod#22]
It'd be nice to also stream these to the monitor page. Perhaps as webp streams?

**Comment by ysmood** [comment]
Yes, I agree. I tried it on my local, not easy. Most solutions have to use FFmpeg, I need to make sure it worth adding an extra heavy dependency.

**Comment by ysmood** [comment]
It will be great if someone who is familiar with streaming can help me.

**Comment by tmc** [comment]
I did some research and it looks like this would cause a dependency on libvpx which would make it not-pure-go..

**Comment by ysmood** [comment]
I usually use Rod on local chrome first, then use the monitor to debug when switching the same code to docker. So I don't have a strong motivation to use it.

But I still think it will be great to have casting supported.

FYI, this is the API: https://chromedevtools.github.io/devtools-protocol/tot/Page/#event-screencastFrame

As you can see, the cast API is still experimental. The best would be chrome team themself to support streaming webp directly so that we don't have to do the heavy lifting. Since chrome itself already have codec lib built-in for web videos. Maybe we should raise a ticket for the chrome team.

**Comment by muhibbudins** [comment]
Hi @ysmood 

Thanks in advance for this great project, and then I've come up with a solution to this issue. Of course, in the way you mentioned earlier, using the built-in PageScreencastFrame (experimental) method of chromium itself.

And for everyone who is also figuring out how to make a video of the running process, you can follow what I have created below.

Create a listener for the screencast frame event

https://github.com/TheOwlEngine/Engine/blob/80144cd25a3abf400b31ce9ca106e9f86a89a56a/main.go#L239-L250

Trigger page screencast frame

https://github.com/TheOwlEngine/Engine/blob/80144cd25a3abf400b31ce9ca106e9f86a89a56a/main.go#L253-L260

Then stop the screencast when the page is closed

https://github.com/TheOwlEngine/Engine/blob/80144cd25a3abf400b31ce9ca106e9f86a89a56a/main.go#L296-L297

Finally, combine each frame into a viewable video

https://github.com/TheOwlEngine/Engine/blob/80144cd25a3abf400b31ce9ca106e9f86a89a56a/main.go#L764-L808

You can see the full source code of my project [here](https://github.com/TheOwlEngine/Engine/blob/main/main.go)

Thank you

**Comment by ysmood** [comment]
@muhibbudins It surprised me that it doesn't require libs like FFmpeg, well done!

I wonder if you could spend some time adding this feature to Rod. I checked the standard of MJPEG, seems like chrome support to stream play it, so maybe we don't have to create temp files, we can stream it directly to a webpage, an example project is here: https://github.com/nsmith5/mjpeg

**Comment by muhibbudins** [comment]
@ysmood Ah, so I can be simpler. I will try to learn about the core of the rod first. Hopefully, I can make a PR for this feature soon

**Comment by muhibbudins** [comment]
@ysmood I already made PR #614 614 for this, but sorry in advance, I didn't understand how `ServeMonitor` works. So, I make this feature work like we use the `Screenshot` function

**Comment by ysmood** [comment]
@muhibbudins check this example https://github.com/go-rod/rod/blob/master/lib/examples/launch-managed/main.go

**Comment by muhibbudins** [comment]
I've seen the code, I mean I haven't caught on why we have to move the video creation process through that serve monitor? whereas we can directly send it into the file directly

**Comment by ysmood** [comment]
When we launch a browser from a remote machine (remote docker cluster), we usually need a way to debug it, being able to see the page live is what `ServeMonitor` wants to achieve.

For example, if you find out the remote scraper gets stuck on a page, you can use this `ServeMonitor` to watch the page and find out the page is not rendered as expected and that's the root cause of the stuck.

**Comment by ysmood** [comment]
We need to use the [<video>](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/video) tag in this [page](https://github.com/go-rod/rod/blob/d726a2af5143012a1feabcba9edd8f90ef117ea6/lib/assets/monitor.html) to play the MJPEG stream of the page.

**Comment by muhibbudins** [comment]
I see, so we just need to add one more function besides the one I've created to monitor the streaming process, right?

**Comment by ysmood** [comment]
Yes, should be very easy, and you even don't need dependencies like `github.com/icza/mjpeg`. Since the protocol of MJPEG is dead simple.

**Comment by muhibbudins** [comment]
Hmmm but what if the end-user wants to take the result stream into an output file? because the main purpose I made this function is for that, so I can share the process of the headless browser with the user.

**Comment by ysmood** [comment]
Create a file with the extension `.mjpeg`, the binary format is simple, just the concatenation of jpeg files. You don't have to use `github.com/icza/mjpeg`, you can just use a browser to play the `.mjpeg` file.

**Comment by muhibbudins** [comment]
@ysmood sorry I'm having an issue where not all frames are sent when I run it in the ServeMonitor function, and it causes the JPEG motion not to run, even though I added a delay when switching pages.

```go
func TestMonitor(t *testing.T) {
  g := setup(t)

  b := rod.New().MustConnect()

  b.Context(g.Context()).ServeMonitor("127.0.0.1:3333")

  page := b.MustPage(g.blank()).MustWaitLoad()

  time.Sleep(5 * time.Second)

  page.Navigate("https://github.com")

  time.Sleep(5 * time.Second)

  page.Navigate("https://google.com")
}
```

And I have tried several ways to decode the data value of the event page screencast frame, but no image is sent to the client.

```go
mux.HandleFunc("/screencast/", func(w http.ResponseWriter, r *http.Request) {
  id := r.URL.Path[strings.LastIndex(r.URL.Path, "/")+1:]
  target := proto.TargetTargetID(id)
  p := b.MustPageFromTargetID(target)

  JPEGQuality := int(90)
  framePerSecond := int(10)

  proto.PageStartScreencast{
    Format:    "jpeg",
    Quality:     &JPEGQuality,
    EveryNthFrame: &framePerSecond,
  }.Call(p)

  flusher, ok := w.(http.Flusher)

  if !ok {
     http.NotFound(w, r)
     return
  }

  w.Header().Add("Content-Type", "multipart/x-mixed-replace; boundary=frame")

  for msg := range p.Event() {
    if msg.Method == "Page.screencastFrame" {
      // first
      w.Write([]byte("\r\n--frame\r\n"))
      w.Write([]byte("Content-Type: image/jpeg\r\nContent-Length: " + strconv.Itoa(len(msg.data)) + "\r\n\r\n"))
      w.Write(msg.data)
      
      // second
      image, _, _ := image.Decode(bytes.NewReader(msg.data))
      jpeg.Encode(w, image, nil)

      io.WriteString(w, "\r\n")
      flusher.Flush()
    }
  }
})
```

I've also tried to follow this [tutorial](https://ivarsblog.com/posts/ssr-stream/) but it turns out to be quite time consuming, even though it looks easy but it's not for me right now.

Thanks

**Comment by ysmood** [comment]
> not all frames are sent

Do you mean the `screencastFrame` doesn't trigger?

**Comment by ysmood** [comment]
A related project: https://github.com/navicstein/rod-stream

cc @navicstein

**Comment by supaplextor** [comment]
I forked an Android adb go package and I think you could take into account adjacent tools like VM videos from virtual box or Android via adb. I'll scrounge some notes up sooner or later. 

Also Apple has a open source streaming media server. I haven't dabbled with it in eons so I'm not sure if it would apply here.

Also my fork solved a few glitches but it's still alpha. After I wrap my head around using events like click on last screen shot, do blah blah blah but that's out of scope here. Only that adb can record the minute videos on internal storage then fetch the video.
